pypga/modules/migen/spimachine.py

```diff
@@ -1,250 +1,328 @@
+        
+        # Counter to count number of (half) cycles
+        self.counter = Signal(max(get_length(data_width),
+                                  get_length(leading_halfcycles),
+                                  get_length(trailing_halfcycles),
+                                  get_length(idle_halfcycles)))
 
+        self.submodules.cg = cg = ClockGen(clock_div)
         self.submodules.reg = reg = Register(data_width)
-        self.submodules.cg = cg = ClockGen(div_width)
-
+        
         ###
-
-        # Bit counter
-        n = Signal.like(self.length, reset_less=True)
-        # Capture end for the in-flight transfer
-        end = Signal(reset_less=True)
-
-        self.comb += [
-                self.ce.eq(cg.done & cg.count)
-        ]
-        self.sync += [
-                If(reg.load,
-                    n.eq(self.length - 1),
-                    end.eq(self.end)
-                ),
-                If(reg.shift,
-                    n.eq(n - 1)
-                )
-        ]
-
+        
         fsm = FSM("IDLE")
         self.submodules += fsm
+        # TODO: Transitions between the FSM states consume one system clock
+        # cycle, such that actually one half cycle is not div but div+1 system
+        # clock cycles long.
+        # Transitions from/to leading, trailing and idle half cycles consume
+        # even one system clock cycle more.
+        # For now, this is not relevant, but for high-speed SPI transmissions
+        # this may play a role.
 
+        # Idle state. Start SPI transaction, when self.start is asserted.
         fsm.act("IDLE",
-                self.idle.eq(1),
-                self.writable.eq(1),
-                self.cs_next.eq(0),#1
-                If(self.load,
-                    cg.count.eq(1),
-                    NextState("START1")
-                )
-        )
-        fsm.act("START1",
-                self.cs_next.eq(1),
-                cg.count.eq(1),
-                cg.extend.eq(1),
-                self.clk_next.eq(0),
-                If(cg.done,
-                   NextState("START2")
-                )
-        )
-        fsm.act("START2",
-                self.cs_next.eq(1),
-                cg.count.eq(1),
-                cg.extend.eq(1),
-                self.clk_next.eq(0),
-                If(cg.done,
-                    If(self.clk_phase,
-                        NextState("PRE"),
-                    ).Else(
-                        cg.extend.eq(1),
-                        reg.load.eq(1),
-                        NextState("SETUP"),
-                    )
-                )
+            self.idle.eq(1),
+            If(self.start,
+                If(self.clk_phase,
+                    # If SPI clk phase is true, there must be at least one
+                    # leading halfcycle. Start the clock generator and set
+                    # counter to 1 in case leading_halfcycles is zero.
+                    NextValue(cg.start, 1),
```

Comment:
I would rather have `cg.start.eq(1)` here (and equivalently in other places), but if I do, then Vivado complains that `start` "has multiple drivers". I don't understand why this happens, since always only one branch should be active. But it works with this sync assignment, which is certainly fine for now.

Reply:
I think the issue lies in the FSM implementation by migen. If you check further down, there is `NextValue(cg.start)`... What did you want to put here? 
```
self.sync += cg.start.eq(1)
```
or 
```
self.comb += cg.start.eq(1)
```
? If it's the latter, the problem is that you cannot both use synchronous statements (NextValue below) and combinartorial logic for the same register.